staging/src/k8s.io/client-go/plugin/pkg/client/auth/azure/azure.go

```diff
@@ -0,0 +1,353 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package azure
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"os"
+	"sync"
+
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/Azure/go-autorest/autorest/azure"
+	"github.com/golang/glog"
+
+	restclient "k8s.io/client-go/rest"
+)
+
+const (
+	azureTokenKey = "azureTokenKey"
+	tokenType     = "Bearer"
+	authHeader    = "Authorization"
+
+	cfgClientID     = "client-id"
+	cfgTenantID     = "tenant-id"
+	cfgAccessToken  = "access-token"
+	cfgRefreshToken = "refresh-token"
+	cfgExpiresIn    = "expires-in"
+	cfgExpiresOn    = "expires-on"
+	cfgEnvironment  = "environment"
+	cfgApiserverID  = "apiserver-id"
+)
+
+func init() {
+	if err := restclient.RegisterAuthProviderPlugin("azure", newAzureAuthProvider); err != nil {
```

Comment:
> I'm a little confused by "client_secret auth". Are you referring to a token endpoint  authentication method or a different oauth2 flow? Can you link to azure docs for this?

@ericchiang Please have a look at page 227 in the develop.pdf document which is attached above. Essentially is an 1-legged flow where a client requests a token for a resource (in this case apiserver) using a secret. This type of token is typically used in service to service authentication.   

> One last nit. Is "azure" the correct name for this plugin? Maybe aadv1?

`azure` should be fine since the SDK is only using the aadv1 endpoints. It's true that this name will reflect more appropriately which AAD version is used. 

@colemickens What's your opinion regarding the name?